File: DAA/C Implementation/Week1/5.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void bubble_sort(int *arr, int n)
{
  if (n == 1)
        return;
    int count = 0;
    for (int i=0; i<n-1; i++)
        {
            if (arr[i] > arr[i+1])
            {
                arr[i]= arr[i] + arr[i+1];
                arr[i + 1] = arr[i] - arr[i+1];
                arr[i] = arr[i] - arr[i+1];
                count++;
            }
        }
      if (count==0)
           return;
    bubble_sort(arr, n-1);
}

int main()
{
    int n = 5000, i;
    clock_t start, end;
    int *arr;

    // Allocating memory dynamically
    arr = (int *)malloc(n * sizeof(int));
    if (arr == NULL)
    {
        printf("Memory allocation failed\n");
        return 0;
    }

    // Best Case: Array is already sorted
    for (i = 0; i < n; i++)
        arr[i] = i;
    start = clock();
    bubble_sort(arr, n);
    end = clock();
    printf("Best Case: %.2f ms\n", ((double)(end - start)) * 1000 / CLOCKS_PER_SEC);

    // Average Case: Array is random
    for (i = 0; i < n; i++)
        arr[i] = rand() % n;
    start = clock();
    bubble_sort(arr, n);
    end = clock();
    printf("Average Case: %.2f ms\n", ((double)(end - start)) * 1000 / CLOCKS_PER_SEC);

    // Worst Case: Array is sorted in reverse order
    for (i = 0; i < n; i++)
        arr[i] = n - i;
    start = clock();
    bubble_sort(arr, n);
    end = clock();
    printf("Worst Case %.2f ms\n", ((double)(end - start)) * 1000 / CLOCKS_PER_SEC);

    free(arr);
    return 0;
}

File: DAA/Week1/5.py
import random
import time

def bubble_sort(arr, n):
    count = 0
    if n == 1:
        return
    for i in range(n - 1):
        if arr[i] > arr[i+1]:
            arr[i], arr[i+1] = arr[i+1], arr[i]
            count = count + 1
    if (count==0):
        return
    bubble_sort(arr, n - 1)

n = 500
arr = []
print("Input size:", n)

# Best Case: Array is already sorted
for i in range(n):
    arr.append(i) 
start = time.time()
bubble_sort(arr,n)
end = time.time()
print("Best Case:", '{:.4f}'.format(end - start), "sec")
arr.clear()

# Average Case: Array is random
for i in range(n):
    arr.append(random.randint(0, n))
start = time.time()
bubble_sort(arr,n)
end = time.time()
print("Average Case:", '{:.4f}'.format(end - start), "sec")
arr.clear()

# Worst Case: Array is sorted in reverse order
for i in range(n):
    arr.append(n-i)
start = time.time()
bubble_sort(arr,n)
end = time.time()
print("Worst Case:", '{:.4f}'.format(end - start), "sec")
arr.clear()